src/Parsers/ExpressionListParsers.cpp

```diff
@@ -599,158 +609,1974 @@ bool ParserLambdaExpression::parseImpl(Pos & pos, ASTPtr & node, Expected & expe
     {
-        if (parseOperator(pos, *it, expected))
-            break;
     }
 
-    /// Let's parse chains of the form `NOT NOT x`. This is hack.
-    /** This is done, because among the unary operators there is only a minus and NOT.
-      * But for a minus the chain of unary operators does not need to be supported.
-      */
-    size_t count = 1;
-    if (it[0] && 0 == strncmp(it[0], "NOT", 3))
+    OperatorType type;
+    Int32 priority;
+    Int32 arity;
+    String function_name;
+};
+
+class Layer
+{
+public:
+    virtual ~Layer() = default;
+
+    bool popOperator(Operator & op)
     {
-        while (true)
-        {
-            const char ** jt;
-            for (jt = operators; *jt; jt += 2)
-                if (parseOperator(pos, *jt, expected))
-                    break;
+        if (operators.empty())
+            return false;
 
-            if (!*jt)
-                break;
+        op = std::move(operators.back());
+        operators.pop_back();
 
-            ++count;
+        return true;
+    }
+
+    void pushOperator(Operator op, bool count = true)
+    {
+        if (count)
+        {
+            if (op.type != OperatorType::Mergeable)
+            {
+                ++depth_diff;
+                ++depth_total;
+            }
+            else
+            {
+                depth_diff -= depth_total;
+                depth_total = 0;
+            }
         }
+
+        operators.push_back(std::move(op));
     }
 
-    ASTPtr elem;
-    if (!elem_parser->parse(pos, elem, expected))
+    bool popOperand(ASTPtr & op)
+    {
+        if (operands.empty())
+            return false;
+
+        op = std::move(operands.back());
+        operands.pop_back();
+
+        return true;
+    }
+
+    void pushOperand(ASTPtr op)
+    {
+        operands.push_back(std::move(op));
+    }
+
+    void pushResult(ASTPtr op)
+    {
+        result.push_back(std::move(op));
+    }
+
+    virtual bool getResult(ASTPtr & op)
+    {
+        if (result.size() == 1)
+        {
+            op = std::move(result[0]);
+            return true;
+        }
+
         return false;
+    }
 
-    if (!*it)
-        node = elem;
-    else
+    virtual bool parse(IParser::Pos & /*pos*/, Expected & /*expected*/, Action & /*action*/)
     {
-        for (size_t i = 0; i < count; ++i)
+        return true;
+    }
+
+    bool isFinished() const
+    {
+        return state == -1;
+    }
+
+    int previousPriority() const
+    {
+        if (operators.empty())
+            return 0;
+
+        return operators.back().priority;
+    }
+
+    OperatorType previousType() const
+    {
+        if (operators.empty())
+            return OperatorType::None;
+
+        return operators.back().type;
+    }
+
+    int empty() const
+    {
+        return operators.empty() && operands.empty();
+    }
+
+    bool lastNOperands(ASTs & asts, size_t n)
```

Comment:
Probably, it's `popLastNOperands`.
Also, probably you can return `std::optional<ASTs>` (but currently it's ok).